aiidalab_widgets_base/computational_resources.py

```diff
@@ -375,126 +437,131 @@ def _write_ssh_config(self, private_key_abs_fname=None):
-                        break
-
-                    elif index == 3:  # Adding a new host.
-                        child.sendline("yes")
-
-                    elif index == 4:
-                        print(
-                            "Failed\nLooks like the key pair is not present in ~/.ssh folder."
-                        )
-                        break
-
-                    elif index == 5:
-                        print("Failed\nUnknown hostname.")
-                        break
-
-                    elif index == 6:
-                        print("Failed\nConnection refused.")
-                        break
+        # If hostname & username are not provided - do not do anything.
+        if self.hostname.value == "":  # check hostname
+            self.message = "Please specify the computer hostname."
+            return False
 
-                    else:
-                        print("Failed\nUnknown problem.")
-                        print(child.before, child.after)
-                        break
-                child.close()
-                yield
+        if self.username.value == "":  # check username
+            self.message = "Please specify your SSH username."
+            return False
 
+        private_key_abs_fname = None
+        if self._verification_mode.value == "private_key":
+            # unwrap private key file and setting temporary private_key content
+            private_key_abs_fname, private_key_content = self._private_key
+            if private_key_abs_fname is None:  # check private key file
+                self.message = "Please upload your private key file."
+                return False
+
+            # Write private key in ~/.ssh/ and use the name of upload file,
+            # if exist, generate random string and append to filename then override current name.
+            self._add_private_key(private_key_abs_fname, private_key_content)
+
+        if not self._is_in_config():
+            self._write_ssh_config(private_key_abs_fname=private_key_abs_fname)
+
+        # Copy public key on the remote computer.
+        ssh_connection_thread = threading.Thread(target=self._ssh_copy_id)
+        ssh_connection_thread.start()
+
+    def _ssh_copy_id(self):
+        """Run the ssh-copy-id command and follow it until it is completed."""
+        timeout = 30
+        self.password_message = f"Sending public key to {self.hostname.value}... "
+        self._ssh_connection_process = pexpect.spawn(
+            f"ssh-copy-id {self.hostname.value}"
+        )
+        while True:
             try:
-                ssh_copy_id()
-            except StopIteration:
-                print(f"Unsuccessful attempt to connect to {self.hostname.value}.")
-                return
+                self.ssh_connection_state = self._ssh_connection_process.expect(
+                    self.SSH_POSSIBLE_RESPONSES,
+                    timeout=timeout,
+                )
+            except pexpect.TIMEOUT:
```

Comment:
This means that if the user did not provide a password within 30 seconds, the process times out and one needs to start all over again. That's a bit of a heavy-handed approach for synchronization between the widget threads and this thread, but I think it is acceptable.

I think there are probably better approaches to this, but I would recommend to move forward with this for now.

Reply:
Not true. The `enter_password` state (0) is not is the cases that will break the loop, so it will keep on asking for input. I think this also the main purpose of the 'while True' loop.

Reply:
The `expect()` function waits on any of the responses in the `SshConnectionState` enum, that includes the password request. The loop is broken in two cases:
1. The `expect()` function times out.
2. We encountered one of the "terminal states" listed below the `try: ... except: ...` block.

This means that once we encountered the "enter_password" state, the loop will continue and wait for the next state. In case that the user does not enter the password within the timeout period (30s), the `pexpect.TIMEOUT` will be raised which triggers the disabling of the password field and a break out of the loop.

Reply:
Yes, I see your point now. Let's keep it 30 s, I think it should be enough for inputting a password.